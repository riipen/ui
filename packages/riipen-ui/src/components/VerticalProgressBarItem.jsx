import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import clsx from "clsx";
import React from "react";
import PropTypes from "prop-types";
import css from "styled-jsx/css";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Grid from "./Grid";
import GridItem from "./GridItem";

const VerticalProgressBarItem = ({ bar, children, classes, color, icon }) => {
  const theme = React.useContext(ThemeContext);

  const getLinkedStyles = () => {
    return css.resolve`
      .icon {
        color: ${theme.palette.common.white};
        z-index: ${theme.zIndex.low};
      }
    `;
  };

  const linkedStyles = getLinkedStyles();

  return (
    <div
      className={clsx(
        "progress",
        color && bar && `progress-${color}`,
        !bar && "progress-none",
        classes
      )}
    >
      <Grid classes={[linkedStyles.className, "wrapper"]} spacing={0}>
        <GridItem lg="auto">
          <div className={clsx("iconWrapper", color && `${color}`)}>
            {!!icon && (
              <FontAwesomeIcon
                className={clsx(linkedStyles.className, "icon")}
                size="sm"
                icon={icon}
              />
            )}
          </div>
        </GridItem>
        <GridItem flexGrow={1} flexShrink={1} lg={10}>
          {children}
        </GridItem>
      </Grid>
      <style jsx>{`
        .default {
          background-color: ${theme.palette.grey[300]};
        }

        .positive {
          background-color: ${theme.palette.positive.main};
        }

        .warning {
          background-color: ${theme.palette.warning.main};
        }

        .progress {
          border-left: 8px solid var(--border-color);
          margin-left: ${theme.spacing(3)}px;
        }

        /* Progress bar colors. */
        .progress-default {
          --border-color: ${theme.palette.grey[300]};
        }

        .progress-none {
          --border-color: transparent;
        }

        .progress-positive {
          --border-color: ${theme.palette.positive.main};
        }

        .progress-warning {
          --border-color: ${theme.palette.warning.main};
        }

        .iconWrapper {
          align-items: center;
          border-radius: 50%;
          display: flex;
          height: 24px;
          justify-content: center;
          margin-bottom: ${theme.spacing(7)}px;
          margin-left: -16px;
          margin-right: ${theme.spacing(5)}px;
          margin-top: -1px;
          width: 24px;
        }
      `}</style>
      {linkedStyles.styles}
    </div>
  );
};

VerticalProgressBarItem.propTypes = {
  // external

  /**
   * The content to render inside.
   */
  children: PropTypes.node,

  /**
   * Additional classes to style with.
   */
  classes: PropTypes.array,

  /**
   * The colour of the progress icon.
   */
  color: PropTypes.oneOf(["positive", "warning", "default"]),

  /**
   * The progress icon.
   */
  icon: PropTypes.object,

  /**
   * Whether to not have bar.
   */
  bar: PropTypes.bool
};

VerticalProgressBarItem.defaultProps = {
  bar: true,
  classes: []
};

export default withClasses(VerticalProgressBarItem);
