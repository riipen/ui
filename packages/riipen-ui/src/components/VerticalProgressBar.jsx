import React from "react";
import PropTypes from "prop-types";

import withClasses from "../utils/withClasses";

import Grid from "./Grid";
import GridItem from "./GridItem";
import VerticalProgressBarItem from "./VerticalProgressBarItem";

const VerticalProgressBar = ({ progresses }) => {
  const renderProgress = () => {
    return progresses.map((progress, index) => (
      <GridItem key={`progress-${index}`}>
        <VerticalProgressBarItem
          progress={{ ...progress, noBar: index >= progresses.length - 1 }}
        />
      </GridItem>
    ));
  };
  return (
    <div>
      <Grid spacing={0}>{renderProgress()}</Grid>
    </div>
  );
};

VerticalProgressBar.propTypes = {
  // external

  /**
   * The progresses to display.
   */
  progresses: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The progress icon.
       */
      icon: PropTypes.object,

      /**
       * The content to render inside.
       */
      children: PropTypes.node,

      /**
       * The colour of the progress icon.
       */
      color: PropTypes.string,

      /**
       * Whether to not have bar.
       */
      noBar: PropTypes.bool
    })
  )
};

VerticalProgressBar.defaultProps = {
  progresses: []
};

export default withClasses(VerticalProgressBar);
