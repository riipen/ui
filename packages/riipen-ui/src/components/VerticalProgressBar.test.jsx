import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import VerticalProgressBar from "./VerticalProgressBar";
import VerticalProgressBarItem from "./VerticalProgressBarItem";

describe("<VerticalProgressBar>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<VerticalProgressBar />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<VerticalProgressBar />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <span>this is a child</span>;

      const wrapper = mount(<VerticalProgressBar>{child}</VerticalProgressBar>);

      expect(wrapper.text()).toContain("this is a child");
    });

    it("passes bar = false to last child", () => {
      const children = [
        <VerticalProgressBarItem key="1" />,
        <VerticalProgressBarItem key="2" />
      ];

      const wrapper = mount(
        <VerticalProgressBar>{children}</VerticalProgressBar>
      );

      expect(
        wrapper
          .find("VerticalProgressBarItem")
          .at(1)
          .props().bar
      ).toBe(false);
    });

    it("does not pass bar to non-last child", () => {
      const children = [
        <VerticalProgressBarItem key="1" />,
        <VerticalProgressBarItem key="2" />
      ];

      const wrapper = mount(
        <VerticalProgressBar>{children}</VerticalProgressBar>
      );

      expect(
        wrapper
          .find("VerticalProgressBarItem")
          .at(0)
          .props().bar
      ).toBe(true);
    });
  });

  describe("classes prop", () => {
    it("applies classes to the root node of VerticalProgressBar", () => {
      const classes = ["classOne"];

      const wrapper = mount(<VerticalProgressBar classes={classes} />);

      expect(
        wrapper
          .find("VerticalProgressBar")
          .childAt(0)
          .props().classes
      ).toContain(classes[0]);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-verticalprogressbar"];

      const wrapper = mount(<VerticalProgressBar />);

      expect(
        wrapper
          .find("VerticalProgressBar")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<VerticalProgressBar classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });
});
