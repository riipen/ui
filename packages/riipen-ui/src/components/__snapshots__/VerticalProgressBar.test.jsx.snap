// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalProgressBar> renders correct snapshot 1`] = `
<withClasses(VerticalProgressBar)>
  <VerticalProgressBar
    classes={
      Array [
        "riipen",
        "riipen-verticalprogressbar",
      ]
    }
  >
    <withClasses(Grid)
      classes={
        Array [
          "riipen",
          "riipen-verticalprogressbar",
        ]
      }
      spacing={0}
    >
      <Grid
        alignItems="flex-start"
        classes={
          Array [
            "riipen",
            "riipen-grid",
            "riipen",
            "riipen-verticalprogressbar",
          ]
        }
        flexDirection="row"
        flexGrow={1}
        flexShrink={1}
        flexWrap="wrap"
        justifyContent="flex-start"
        onResize={[Function]}
        spacing={0}
      >
        <div
          className="jsx-2773517404 riipen riipen-grid riipen riipen-verticalprogressbar md"
        />
        <JSXStyle
          dynamic={
            Array [
              "flex-start",
              "row",
              1,
              1,
              "wrap",
              "flex-start",
              0,
              0,
            ]
          }
          id="3466573101"
        />
      </Grid>
    </withClasses(Grid)>
  </VerticalProgressBar>
</withClasses(VerticalProgressBar)>
`;
